File: learning/submitted_projects_CS50x/credit/credit - Copy.py
from cs50 import get_int

def iv():
    print("INVALID")

card = get_int("Card: ")
cal = card
odd_sum = 0
even_sum = 0

# apply Luhn’s algorithm
while cal > 0:
    odd_sum += cal % 10
    cal //= 10
    if cal == 0:
        break
    even_di = cal % 10
    even_di = even_di * 2
    if even_di > 9:
        even_di -= 9
    even_sum += even_di
    cal //= 10

k = even_sum + odd_sum
length = len(str(card))

if k % 10 == 0:
    if length == 15:
        prefix = card // 10**13
        if prefix in (34, 37):
            print("AMEX")
        else:
            iv()
    elif length == 16:
        prefix2 = card // 10**14
        prefix1 = card // 10**15
        if prefix2 in range(51, 56):
            print("MASTERCARD")
        elif prefix1 == 4:
            print("VISA")
        else:
            iv()
    elif length == 13 and card // 10**12 == 4:
        print("VISA")
    else:
        iv()
else:
    iv()

File: learning/submitted_projects_CS50x/credit/credit - Copy.c
#include <cs50.h>
#include <stdio.h>
int getValue1(int x);
int getValue2(int x);
// name a fuction then put it
typedef void (*FunctionPointer)();
void repeat(int n, FunctionPointer func);
void check();
int codi(long n);
void iv();

long n;
int z;
long cal;
int sum1 = 0;
int sum2 = 0;

int main()
{


    n = get_long("Number: ");
    z = codi(n);
    cal = n;
    repeat(8, check);
    int k = sum1 + sum2;

    // first check that the sum is divisable by 10
    if ((k % 10) == 0)
    {
        // this is for amex
        if (z == 15)
        {
            if (n / 10000000000000 == 34 || n / 10000000000000 == 37)
            {
                printf("AMEX\n");
            }
            else
            {
                iv();
            }
        }

        // this is for visa
        else if (z == 16 || z == 13)
        {
            if ((n / 1000000000000000 == 4) || (n / 1000000000000 == 4))
            {
                printf("VISA\n");
            }

            // this  is for mastercard
            else if (n / 100000000000000 > 50 && n / 100000000000000 < 56)
            {
                printf("MASTERCARD\n");
            }

            else
            {
                iv();
            }
        }
        else
        {
            iv();
        }

    }

    else
    {
        iv();
    }
}

int getValue2(int x)
{
    if (x >= 5)
    {
        x = x * 2 - 9;
        return x;
    }
    else
    {
        return x * 2;
    }
}

int getValue1(int x)
{
    return x;
}

// the first gap put the amount of number
// the second gap put the name of the function
void repeat(int x, FunctionPointer func)
{
    for (int i = 0; i < x; i++)
    {
        func();
    }
}

void check()
{
    int y = cal % 10;
    sum1 += getValue1(y);
    cal = cal / 10;
    int x = cal % 10;
    sum2 += getValue2(x);
    cal = cal / 10;
}

int codi(long x)
{
    int count = 0;
    while (x > 0)
    {
        count++;
        x /= 10;
    }
    return count;
}

void iv()
{
    printf("INVALID\n");
}
